runtime/Python2/src/antlr4/Recognizer.py

```diff
@@ -45,14 +45,20 @@ def removeErrorListener(self, listener):
     def removeErrorListeners(self):
         self._listeners = []
 
+    def getTokenNames(self):
+        return tuple(self.literalNames + self.symbolicNames[len(self.literalNames):])
```

Comment:
why return a tuple rather than a list ?

Reply:
This is being used as a dict key in the calling function and so needs to be a hashable type, of which list is not one; the alternative is to cast it to a tuple in getTokenTypeMap/getRuleIndexMap. Would you prefer that?

Reply:
a collection of token names used as a dict key ? not sure I understand...

Reply:
also I have some doubt about your logic. The name of a token can be its literal name or its symbolic name, concatenating both lists makes no sense to me.

Reply:
have you tested this ?

Reply:
1.   Yes, the implementations of getTokenTypeMap/getRuleIndexMap use the collection as a dict key, see lines 69 and 72: https://github.com/antlr/antlr4/blob/76fa05c21b12b96a6c12a0a82e611ed9d87d5af4/runtime/Python2/src/antlr4/Recognizer.py#L64-L73
2.   Literals and symbols are two different types of tokens, none have both a literal and a symbolic name. Taking from the grammar I've been working on, the lexer's ruleNames are like `[ "T__0", "T__1", ... "AND", "OR", "NOT" ]`, and these correspond to the properties on the generated Lexer class, where `T__0 = 1`, `AND = 20`, etc. The lexer's symbolic names are only "AND", etc, and not "T__0" etc, while the literal names are `"'('"` etc. So this concatenation would generate a list that can be used like `tokenNames[20]` produces `"AND"` and `tokenNames[1]` produces `"'('"`, with `"T__0"` etc not in the list at all.
    Now, there is a difference between the generated Lexer and Parser classes that I hadn't noticed before and I will have to fix; the Parser class's `symbolicNames` contains the string `"<INVALID>"` for index 0 and all the literal token types, so my first symbol "AND" is at index 20. But the Lexer class has one `"<INVALID>"` at index 0 and then "AND" is at index 1.
3.    Only manually, with the Parser subclass, and I ran the existing parse tests for the Python targets. I couldn't find a convenient way to test any other functionality specific to the Python runtimes.